src/libcore/cell.rs

```diff
@@ -1022,12 +1021,12 @@ impl<'b> BorrowRef<'b> {
     #[inline]
     fn new(borrow: &'b Cell<BorrowFlag>) -> Option<BorrowRef<'b>> {
         let b = borrow.get();
-        if b >= MIN_WRITING {
+        if b < UNUSED {
             None
         } else {
             // Prevent the borrow counter from overflowing into
             // a writing borrow.
-            assert!(b < MIN_WRITING - 1);
+            assert!(b != isize::max_value());
```

Comment:
This assert was new in #51466 -- I missed the chance to comment there, but AFAICT (contrary to https://github.com/rust-lang/rust/pull/51466/files#r194758679), returning `None` here is exactly what the code did previously (before #51466). So I want to bring it up again: why do this? If a `mem::forget(rc.borrow())` loop overflows the borrow flag into the "writing" range, we'll just errenously consider the RefCell mutably borrowed and refuse to create any new Refs or RefMuts, even without any extra checks. That's not quite as tidy as this assertion firing, but it's sound and the diagnostic value of this assertion failing isn't very high anyway. And since we're optimizing anyway...

Reply:
That's a fair point. Given that there are no `RefMut`s in existence, there'd be no way to `RefMut::map_split` in order to create more `RefMut`s.

I suppose my one concern would be about the wrapping semantics. Is it well-defined that, in release mode, overflows wrap around to `isize::min_value()`? If not, we'll want explicit wrapping addition.

Reply:
Wow, this is a crazy trick. If panicking is the problem, why not turn the `assert!` into
```
if b == isize::max_value() { return None; }
```
That still keeps this function panic-free but avoids this strange state where the borrow flag is negative but there are `Ref` around.

Reply:
I don't know if panicking is the problem, everything's complete speculation anyway, extra computation and branching might also be an issue.

I could have sworn that the "benign overflow" trick is pre-existing (before #51466 that is). But on some more speluking, it seems it was just discussed, never actually implemented? We should probably try matching the previous behavior (i.e., returning None) first.

Reply:
> We should probably try matching the previous behavior (i.e., returning None) first.

I would prefer that, also for the sake of the sanity of whoever will try to prove this correct. ;)

Reply:
Ah, I missed this comment. If that's the preference, I'll re-evaluate the discussion below about assertions.